src/poly.jl

```diff
@@ -83,32 +83,30 @@ Similar to `p[i]` but instead of an `Arb` or `Acb` returns an `ArbRef`
 or `AcbRef` which still shares the memory with the `i`-th entry of
 `p[i]`.
 
-For `ArbPoly` and `AcbPoly` it only allows accessing coefficients up
-to the degree of the polynomial since higher coefficients are not
-guaranteed to be allocated.
-
-For `ArbSeries` and `AcbSeries` it allows accessing coefficients up to
-the degree of the series. Note that this degree might not be the same
-as the degree of the underlying polynomial (in case higher order
-coefficients are zero) but the way they are constructed ensures that
-the coefficients will always be initialised.
+It only allows accessing coefficients that are allocated. For
+`ArbPoly` and `AcbPoly` this is typically the degree of the
+polynomial, but can be higher if for example `Arblib.fit_length!` is
+used. For `ArbSeries` and `AcbSeries` all coefficients up to the
+degree of the series are guaranteed to be allocated, even if the
+underlying polynomial has a lower degree.
 
 !!! Note: If you use this to change the coefficient in a way so that
     the degree of the polynomial might change you need to normalise
     the polynomial afterwards to make sure that Arb recognises the
     possibly new degree of the polynomial. If the new degree is the
-    same or lower this can be done using `Arblib.normalise!`. If the
-    new degree is higher you need to manually set the correct value of
-    `Arblib.cstruct(p).length` to be one higher than the new degree.
+    same or lower this can be done using `Arblib.normalise!(p)`. If
+    the new degree is higher you need to manually set the length with
+    `Arblib.set_length!(p, len)`, where `len` is one higher than the
+    new degree.
```

Comment:
this sounds a bit hacky (unsafe?) Not that I know a better way of doing it ;) to the contrary, I barely understand what is this about ;) still I see that there's a possibility of entering into the situation of inconsitent/garbage data.

Maybe `ref` should allow the access only for up to `degree` and `unsafe_ref` for the allocated coeffs?

Reply:
Yes, I don't like it very much and I have indeed shot myself in the foot a few times with it. However I have yet to find a better solution.

For `ArbPoly` and `AcbPoly` it would be fine to only allow access up to `degree`, if you then really need to access higher coefficients to set them you would call `Arblib.set_length!` first. Note that there is still the problem that you need to `normalise!` if you set the top coefficient to zero. In theory you could also run into issues if you take the reference, then mutate the polynomial in a way that changes the degree and then try to use the reference, that would be equivalent to taking a reference above the degree to begin with.

For `ArbSeries` and `AcbSeries` it is more complicated. In this case the degree of the series is not necessarily the same as the degree of the underlying polynomial. If you only allow access to the degree of the underlying polynomial then many functions get annoying to implement because you will need to have a special case for when the underlying degree is lower than the coefficient you are trying to access (something that happens fairly often).

To conclude I don't think it is a good idea to change into `ref` and `unsafe_ref`, because the `ref` will always be unsafe in some ways and the naming might give you the impression that it is not.

I have updated the documentation to hopefully be a bit more clear. Though I'm not sure if the example I gave is the best.

Reply:
I realized I was not happy with the example I have so I made a new version. It now gives one explicit example of issues when the degree is lowered and another when the degree is increased. Since it is fairly long I put it in the extended help.